torch/distributed/_sharded_tensor/api.py

```diff
@@ -214,6 +223,76 @@ def _init_rpc(self):
         # Barrier for all RPCs to finish on all ranks.
         rpc.api._barrier(rpc_workers)
 
+    @classmethod
+    def init_from_local_shards(
+        cls,
+        local_shards: List[torch.Tensor],
+        shards_metadata: List[ShardMetadata],
+        process_group=None
+    ):
+
+        if len(local_shards) == 0 or len(shards_metadata) == 0:
+            raise ValueError("local_shards or shards_metadata must not be empty!")
+
+        sharded_tensor = cls.__new__(cls)
+
+        # prepare initialization
+        sharded_tensor._prepare_init(process_group=process_group)
+
+        sharded_tensor._sharding_spec = EnumerableShardingSpec(shards_metadata)
+        sharded_tensor._sharding_metadata = shards_metadata
+
+        current_rank = dist.get_rank(sharded_tensor._process_group)
+
+        local_shard_idx = 0
+        num_dims = len(shards_metadata[0].shard_lengths)
+        offset_length_tuple_list_by_dim: List[List[Tuple[int, int]]] = [[] for _ in range(num_dims)]
+        # validate the metadata and add to local_shards
+        for shard_metadata in shards_metadata:  # type: ignore[attr-defined]
+            rank, local_device = sharded_tensor._parse_and_validate_remote_device(shard_metadata.placement)
+            # record shard_offsets/lengths for dims recovery later
+            shard_offsets = shard_metadata.shard_offsets
+            shard_lengths = shard_metadata.shard_lengths
+
+            for i, (offset, length) in enumerate(zip(shard_offsets, shard_lengths)):
+                offset_length_tuple_list_by_dim[i].append((offset, length))
+
+            # validate local_shards and metadata
+            if current_rank == rank:
+                local_shard = local_shards[local_shard_idx]
+                if shard_metadata.shard_lengths != list(local_shard.size()):
```

Comment:
Isn't this assuming some sort of order between shard_metadata and local_shards? For example doesn't this fail if the first shard_metadata we encounter is for local_shards[1]? Local shards could have different shard_lengths.

Reply:
yep here we are assuming the order passed in between `local_shards` and `shards_metadata` are consistent, this is a requirement from the user, which might be not that nice, but we have to specify the order in order for the corresponding metadata and local shard to match, I couldn't think of a better idea rather than pushing this requirement to the user yet, do you have any insights on how we could match them seamlessly without imposing the order?

Reply:
The shard_metadata contains the location of all shards, based on that you can figure out which metadata belongs to this rank and then validate that with local_shards?

Reply:
Yeah we can do that, the ordering of metadata won't matter across locations as we can figure the location out from shard_metadata and assert the local tensor is on the same location. But the problem still happens if we have multiple tensors in the same location, consider only one location `cuda:0`, suppose there're two shards/tensors for location `cuda:0`, there're two shard_metadatas associated with them, and if they are the same size, how do we know which tensor match to which shard_metadata? The current approach I am doing for this case is, since the shard_metadata passed in is a list, the first occuring shard_metadata that has location `cuda:0` should match to the first tensor,